env/io_posix.cc

```diff
@@ -663,82 +664,81 @@ IOStatus PosixRandomAccessFile::MultiRead(FSReadRequest* reqs,
         rep_to_submit = &req_wraps[reqs_off++];
       }
       assert(rep_to_submit->req->len > rep_to_submit->finished_len);
-      rep_to_submit->iov.iov_base =
+      rep_to_submit->buf =
           rep_to_submit->req->scratch + rep_to_submit->finished_len;
-      rep_to_submit->iov.iov_len =
+      rep_to_submit->len =
           rep_to_submit->req->len - rep_to_submit->finished_len;
 
       struct io_uring_sqe* sqe;
       sqe = io_uring_get_sqe(iu);
-      io_uring_prep_readv(
-          sqe, fd_, &rep_to_submit->iov, 1,
+      io_uring_prep_read(
+          sqe, fd_, rep_to_submit->buf, rep_to_submit->len,
           rep_to_submit->req->offset + rep_to_submit->finished_len);
       io_uring_sqe_set_data(sqe, rep_to_submit);
+      io_uring_submit(iu);
     }
     incomplete_rq_list.clear();
 
-    ssize_t ret =
-        io_uring_submit_and_wait(iu, static_cast<unsigned int>(this_reqs));
-    if (static_cast<size_t>(ret) != this_reqs) {
-      fprintf(stderr, "ret = %ld this_reqs: %ld\n", (long)ret, (long)this_reqs);
-    }
-    assert(static_cast<size_t>(ret) == this_reqs);
+    size_t cqes_left = this_reqs;
+    while (cqes_left > 0) {
+      struct io_uring_cqe* cqes[cqes_left];
+      size_t completed = io_uring_peek_batch_cqe(iu, cqes, cqes_left);
```

Comment:
IIUC, ```io_uring_peek_batch_cqe``` doesn't wait for any completions, right? So is this effectively a busy wait loop?

Reply:
Yep, correct :)

Reply:
Can we check ```allow_io_polling``` here and , if false, use the wait variant?

Reply:
Yeah, definitely. But it shouldn't be necessary - since `io_uring_submit_and_wait()` is called in case `allow_io_polling` is set, all IOs will be ready to be fetched when we call `io_uring_peek_batch_cqe()` here

Reply:
Just pinging you @anand1976 - the thread was marked as resolved when I commented, so I'm not sure if you got a notification.